### PR TITLE
Unread marker takes hidden messages into account.

### LINKED ISSUE #604: Unread marker shows up incorrectly when joins/parts are hidden
To reproduce:
1. Hide joins/parts
2. Join a channel
3. Wait for someone to join/part
4. Observe 'New messages' marker appears, despite there being apparently no messages in the channel.

_the_lounge

### PR DESCRIPTION
**Pull request by swordbeta**
Stops showing the unread messages marker when there are hidden joins/parts/quits/nick changes which is rather annoying. Would close #604, went with the solution suggested by xPaw.

### CODE REVIEW

File: client/js/lounge.js [hunk 1]
```diff
@@ -443,7 +443,9 @@ $(function() {
 				data
 			]);
 
-		if (data.msg.self) {
+		var lastVisibleChild = container.find("div:visible:last-child");
+		if (data.msg.self
+			|| (lastVisibleChild.length === 1 && lastVisibleChild.hasClass("unread-marker"))) {
```
Comment (xPaw) [inline client/js/lounge.js]: The `length` check isn't really needed.

File: client/js/lounge.js [hunk 2]
```diff
@@ -443,7 +443,9 @@ $(function() {
 				data
 			]);
 
-		if (data.msg.self) {
+		var lastVisibleChild = container.find("div:visible:last-child");
```
Comment (xPaw) [inline client/js/lounge.js]: I'm wondering what kind of performance such a selector will have.

**Comment by swordbeta** [comment]
@xPaw Removed the length check.

Not sure about performance, according to [this jsperf](https://jsperf.com/jquery-visible-test) `:visible` makes things significantly slower. I wonder if this will actually matter in the lounge though. And according to [this jsperf](https://jsperf.com/last-child-vs-jquery-last) `.last()` is faster than `:last-child`, so I changed that.

File: client/js/lounge.js [hunk 3]
```diff
@@ -443,7 +443,8 @@ $(function() {
 				data
 			]);
 
-		if (data.msg.self) {
+		if (data.msg.self
+			|| container.find("div:visible").last().hasClass("unread-marker")) {
```
Comment (xPaw) [inline client/js/lounge.js]: Also, this might not work if there is a date marker appended in the code above.

Comment (astorije) [inline client/js/lounge.js]: Is it something that an extra check for the date marker could fix? Not sure.

**Comment by astorije** [comment]
@xPaw, mind helping with this? I'm not sure how to go with your comment. @swordbeta, do you need help or just time to amend this?

**Comment by astorije** [comment]
@xPaw, ping?

**Comment by swordbeta** [comment]
@xPaw @astorije Updated, I'm not sure about performance though

**Comment by xPaw** [comment]
Yeah that definitely looks like performance destroyer.

**Comment by swordbeta** [comment]
@xPaw Updated it, any other suggestions that might improve performance?

File: client/js/lounge.js [hunk 4]
```diff
@@ -442,8 +442,11 @@ $(function() {
 				target,
 				data
 			]);
-
-		if (data.msg.self) {
+		var lastVisible = container.find("div:visible").last();
+		if (data.msg.self
+			|| lastVisible.hasClass("unread-marker")
+			|| (lastVisible.hasClass("date-marker")
+				&& lastVisible.prev().hasClass("unread-marker"))) {
```
Comment (xPaw) [inline client/js/lounge.js]: Would this prev() need visible check too? I understand why its necessary to check previous element, but this does not feel like performant code 😞

**Comment by astorije** [comment]
@xPaw, mind reviewing/approving this first before I give it a review?

**Comment by astorije** [comment]
@xPaw, ping?

**Comment by xPaw** [comment]
@astorije I guess we can try merging this, there's really no better way of fixing it I don't think.

**Comment by astorije** [comment]
Thank you for your help on this, @swordbeta!

**Comment by astorije** [comment]
Hey @swordbeta, we have free sticker packs for our contributors now!
If you're interested, go fill the form at https://goo.gl/forms/f5usqAEp5DWoeXC92 🙂

**Pull request merged**